contracts/BondCalculator.sol

```diff
@@ -0,0 +1,301 @@
+    uint32 vesting;
+}
+
+/// @title BondCalculator - Smart contract for bond calculation payout in exchange for OLAS tokens based on dynamic IDF.
+/// @author Aleksandr Kuperman - <aleksandr.kuperman@valory.xyz>
+/// @author Andrey Lebedev - <andrey.lebedev@valory.xyz>
+/// @author Mariapia Moscatiello - <mariapia.moscatiello@valory.xyz>
+contract BondCalculator {
+    event OwnerUpdated(address indexed owner);
+    event DiscountParamsUpdated(DiscountParams newDiscountParams);
+
+    // Maximum sum of discount factor weights
+    uint256 public constant MAX_SUM_WEIGHTS = 10_000;
+    // OLAS contract address
+    address public immutable olas;
+    // Tokenomics contract address
+    address public immutable tokenomics;
+    // veOLAS contract address
+    address public immutable ve;
+
+    // Contract owner
+    address public owner;
+    // Discount params
+    DiscountParams public discountParams;
+
+
+    /// @dev Bond Calculator constructor.
+    /// @param _olas OLAS contract address.
+    /// @param _tokenomics Tokenomics contract address.
+    /// @param _ve veOLAS contract address.
+    /// @param _discountParams Discount factor parameters.
+    constructor(address _olas, address _tokenomics, address _ve, DiscountParams memory _discountParams) {
+        // Check for at least one zero contract address
+        if (_olas == address(0) || _tokenomics == address(0) || _ve == address(0)) {
+            revert ZeroAddress();
+        }
+
+        olas = _olas;
+        tokenomics = _tokenomics;
+        ve = _ve;
+        owner = msg.sender;
+
+        // Check for zero values
+        if (_discountParams.targetNewUnits == 0 || _discountParams.targetVotingPower == 0) {
+            revert ZeroValue();
+        }
+        // Check the sum of factors that cannot exceed the value of 10_000 (100% with a 0.01% step)
+        uint256 sumWeights;
+        for (uint256 i = 0; i < _discountParams.weightFactors.length; ++i) {
+            sumWeights += _discountParams.weightFactors[i];
+        }
+        if (sumWeights > MAX_SUM_WEIGHTS) {
+            revert Overflow(sumWeights, MAX_SUM_WEIGHTS);
+        }
+        discountParams = _discountParams;
+    }
+
+    /// @dev Changes contract owner address.
+    /// @param newOwner Address of a new owner.
+    function changeOwner(address newOwner) external {
+        // Check for the contract ownership
+        if (msg.sender != owner) {
+            revert OwnerOnly(msg.sender, owner);
+        }
+
+        // Check for the zero address
+        if (newOwner == address(0)) {
+            revert ZeroAddress();
+        }
+
+        owner = newOwner;
+        emit OwnerUpdated(newOwner);
+    }
+
+    /// @dev Changed inverse discount factor parameters.
+    /// @param newDiscountParams Struct of new discount parameters.
+    function changeDiscountParams(DiscountParams memory newDiscountParams) external {
```

Comment:
So this is global, not product specific?

Reply:
Yes, as otherwise this would need to be voted on for every product.

Reply:
Technically each product is voted on. So wouldn't it be more useful to be able to set this as part of the product? Just thinking out loud.

Reply:
Considering how products are launched, those would be pretty much the same. Also, if something happens on a protocol level, my understanding is that we want all the bonds to follow new parameters even for unfinished products, since otherwise the old params could become wrong / unjustified.